shared/src/balancer_event_handler.rs

```diff
@@ -54,15 +130,124 @@ impl PoolSpecialization {
+                );
+            }
+            Some(existing_pool) => {
+                // If exists before PoolRegistration, then only pool_id and tokens are known
+                existing_pool.update_from_event(BalancerEvent::PoolRegistered(registration));
+                tracing::debug!(
+                    "Pool Address and specialization recovered for existing Balancer pool with id {:}", existing_pool.pool_id
+                );
+            }
+        }
+    }
+
+    fn insert_token_data(&mut self, index: EventIndex, registration: TokensRegistered) {
+        let pool_id = &registration.pool_id;
+        match self.pools.get_mut(pool_id) {
+            None => {
+                // TokensRegistered event received before PoolRegistered, leaving pool_address
+                // empty until PoolRegistration event is processed.
+                tracing::debug!(
+                    "Creating Balancer Pool with id {:?} from TokensRegistration event (pool address pending...)", pool_id
+                );
+                self.pools
+                    .insert(*pool_id, WeightedPool::from((index, registration.clone())));
+            }
+            Some(existing_pool) => {
+                // If pool exists before token registration, then only pool_id and address known
+                existing_pool
+                    .update_from_event(BalancerEvent::TokensRegistered(registration.clone()));
+                tracing::debug!(
+                    "Pool Tokens recovered for existing Balancer pool with id {:?}",
+                    pool_id
+                )
+            }
+        }
+
+        // In either of the above cases we can now populate pools_by_token
+        for token in registration.tokens {
+            self.pools_by_token
+                .entry(token)
+                .or_default()
+                .insert(*pool_id);
+        }
+    }
+
+    fn replace_events(
+        &mut self,
+        delete_from_block_number: u64,
+        events: Vec<(EventIndex, BalancerEvent)>,
+    ) -> Result<()> {
+        self.delete_pools(delete_from_block_number)?;
+        self.insert_events(events)?;
+        Ok(())
+    }
+
+    fn delete_pools(&mut self, delete_from_block_number: u64) -> Result<()> {
+        let pool_ids_to_delete = self
+            .pools
+            .iter()
+            .filter_map(|(pool_id, pool)| {
+                if pool.block_created >= delete_from_block_number {
+                    return Some(*pool_id);
+                }
+                None
+            })
+            .collect::<HashSet<H256>>();
+
+        self.pools
+            .retain(|pool_id, _| !pool_ids_to_delete.contains(pool_id));
+        // Remove all deleted pool_ids from token listing.
+        // Note that this could result in an empty set for some tokens.
+        for (_, pool_set) in self.pools_by_token.iter_mut() {
+            *pool_set = pool_set
+                .difference(&pool_ids_to_delete)
```

Comment:
Could also just union with the remaining pools (then you wouldn't need the filter_map).

Reply:
Do you mean intersect?

Reply:
I will make this change because it is a complimentary equivalent and removes the negation (`!`), but I don't see a `filter_map` being used anywhere here.